Source/Core/DolphinQt/GBAWidget.cpp

```diff
@@ -0,0 +1,316 @@
+// Copyright 2021 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt/GBAWidget.h"
+
+#include <fmt/format.h>
+
+#include <QAction>
+#include <QCloseEvent>
+#include <QContextMenuEvent>
+#include <QFileDialog>
+#include <QIcon>
+#include <QImage>
+#include <QMenu>
+#include <QPainter>
+
+#include "AudioCommon/AudioCommon.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/Core.h"
+#include "Core/HW/SI/SI.h"
+#include "Core/HW/SI/SI_Device.h"
+#include "Core/Movie.h"
+#include "Core/NetPlayClient.h"
+#include "DolphinQt/Resources.h"
+#include "DolphinQt/Settings.h"
+
+GBAWidget::GBAWidget(int device_number, std::string_view current_rom, std::string_view title,
+                     u32 width, u32 height, QWidget* parent, Qt::WindowFlags flags)
+    : QWidget(parent, flags), m_device_number(device_number), m_geometry_slot(device_number),
+      m_current_rom(current_rom), m_game_title(title), m_width(width), m_height(height),
+      m_volume(0), m_muted(false)
+{
+  bool visible = true;
+
+  SetVolume(100);
+
+  if (NetPlay::IsNetPlayRunning())
+  {
+    auto client = Settings::Instance().GetNetPlayClient();
+    auto pid = client->GetPadMapping()[m_device_number];
+    for (const NetPlay::Player* player : client->GetPlayers())
+    {
+      if (player->pid == pid)
+        m_netplayer_name = player->name;
+    }
+
+    if (client->IsLocalPlayer(pid))
+    {
+      m_geometry_slot = client->InGamePadToLocalPad(m_device_number);
+    }
+    else
+    {
+      int ingame_slot = client->LocalPadToInGamePad(m_device_number);
+      if (ingame_slot < 4)
+        m_geometry_slot = ingame_slot;
+      if (client->GetNetSettings().m_HideRemoteGBAs)
+        visible = false;
+      if (!IsMuted())
```

Comment:
Should remote GBAs be unconditionally muted? I'd either make this depend on `HideRemoteGBAs` (so all visible GBAs play sound and all invisible do not) or even separate into its own option.